nixos/modules/misc/assertions.nix

```diff
@@ -1,34 +1,155 @@
 { lib, ... }:
+{
+  options = {
+    assertions = lib.mkOption {
+      type = let
+        assertionItemType = lib.types.submodule {
+          options = {
+            enable = lib.mkOption {
+              description = ''
+                Whether to enable this assertion.
 
-with lib;
+                This option is mostly useful for users, in order to forcefully disable assertions they believe to be
+                erroneous while waiting for someone to fix the assertion upstream.
+              '';
+              type = lib.types.bool;
+              default = true;
+              example = false;
+            };
 
-{
+            assertion = lib.mkOption {
+              description = "Condition to be asserted. If this is `false`, the evaluation will throw the error";
+              type = lib.types.bool;
+              example = false;
+            };
 
-  options = {
+            message = lib.mkOption {
+              description = "The contents of the error message that should be shown upon triggering a false assertion";
+              type = lib.types.nonEmptyStr;
+              example = "This is an example error message";
+            };
+          };
+        };
 
-    assertions = mkOption {
-      type = types.listOf types.unspecified;
+        # This might be replaced when tagged submodules or similar are available,
+        # see https://github.com/NixOS/nixpkgs/pull/254790
+        checkedAssertionItemType = let
+          check = x: x ? assertion && x ? message;
+        in lib.types.addCheck assertionItemType check;
```

Comment:
You don't need an extra `config` fixpoint, or `options` introspection, or anything like that, so a [record](https://github.com/NixOS/nixpkgs/pull/334680) would be a better type constructor for this, instead of `submodule`. It does the thing you need, and basically no redundant or confusing features.

Reply:
I'm currently using both `config` and `options` for the lazyness toggle in a submodule fixpoint though. I'll have a look at it, but I'm not sure how well it would work with the current solution.